### ISSUE TITLE
TPU train_and_evaluate: nontrivial evaluation_master

### ISSUE THREAD

**Issue by DocIsenberg** [tensorflow/tpu#606]
Hi all,

At a high level, I am trying to setup the following usage pattern:

- Two TPUs, we'll call them tpu-a, and tpu-b, and a compute vm we'll call vm-a

- tpu-a should train my model, tpu-b should evaluate on new checkpoints periodically (and preferably again once at the end, but that's easy to do separately), vm-a manages

I'd like this setup to prevent repeated graph construction/tear down when switching between training/eval modes on the same TPU.

I'd like to take advantage of `train_and_evaluate` to manage checkpoints, setup the eval hooks (using EvalSpec with `start_delay_seconds` and `throttle_seconds`), etc.
I'm new to tensorflow cluster configuration, so I have tried a number of things to get this configuration working. All of them either 1.) Failed to satisfy the goals of the configuration or 2.) failed to evaluate, only train.

Before I list my attempts, here is some relevant code that I use to generate the cluster specifications. If more code would be useful here, I am happy to provide.

Also, TF_CONFIG is unset in my application, I have a note on this below.

```  tpu_cluster_resolver = None
  if FLAGS.use_tpu and FLAGS.tpu_name:
    tpu_cluster_resolver = tf.contrib.cluster_resolver.TPUClusterResolver(
        FLAGS.tpu_name, zone=FLAGS.tpu_zone, project=FLAGS.gcp_project)
    tpu_strategy = tf.contrib.distribute.TPUStrategy(
      tpu_cluster_resolver, steps_per_run=num_train_steps)
    if FLAGS.eval_tpu_name:
      # If supply an eval tpu, construct cluster with both tpus
      eval_job_name = "evaluator"
      eval_tpu_cluster_resolver = tf.contrib.cluster_resolver.TPUClusterResolver(
        FLAGS.eval_tpu_name, zone=FLAGS.tpu_zone, project=FLAGS.gcp_project,
        job_name=eval_job_name)
      # tpu_union_cluster_resolver = tf.contrib.cluster_resolver.UnionClusterResolver(
      #   tpu_cluster_resolver,
      #   eval_tpu_cluster_resolver)
      eval_tpu_strategy = tf.contrib.distribute.TPUStrategy(
        eval_tpu_cluster_resolver, steps_per_run=eval_steps)
      eval_master_address = eval_tpu_cluster_resolver.get_master() #"grpc://" + eval_tpu_cluster_resolver.cluster_spec().task_address(eval_job_name, 0)
    else:
      eval_tpu_strategy = None # TODO: test if "none" or if another tpustrategy identical to train with diff steps
      eval_master_address = None
```

- My initial try was to create two TPUClusterResolvers, and then union them together using UnionClusterResolver, and supply the union to create a `tpu.RunConfig` with `cluster=tpu_union_cluster_resolver`, and then setting `master=FLAGS.master` and `evaluation_master=eval_master_address` favorably. However, this fixes both tpus into the training, requiring the graph to be torn down regardless, so that was a no-go.

- The next attempt was to keep the ClusterResolvers separate, pass only training cluster resolver into `tpu.RunConfig`, as well as the training master, but then set `evaluation_master` to `eval_tpu_resolver.get_master()`. Training would run, but no eval would run.

- I tried leaving out the cluster specification in the `tpu.RunConfig` and instead relying on specifying the clusters via TPUStrategy, similar to what is done here: [distribution_strategy](https://github.com/tensorflow/tpu/blob/master/models/experimental/distribution_strategy/resnet_estimator.py). However, without the `cluster` argument, the operation was not able to discover the TPUs at all.

I fiddled around a bit with each of these solutions, and my research has raised even more questions for myself:

- I'm curious about directly specifying the cluster by setting TF_CONFIG, because much of the documentation I have read ([train_and_evaluate](https://www.tensorflow.org/api_docs/python/tf/estimator/train_and_evaluate)) about non-TPU clusters indicates that cluster configurations should be set that way, and in particular the evaluation node should have task type set to `evaluator` and it should not be present in the training cluster spec. However, I have seen no documentation or examples anywhere about using TF_CONFIG in conjunction with TPUs, and I am unsure of how to ssh into them to set that env variable (there's nothing [here](https://cloud.google.com/sdk/gcloud/reference/beta/compute/tpus/)), or if doing so even makes sense, as I am unsure how the TPU internal containers are configured and spun up.

- I'm curious about what restrictions are on `evaluation_master` in `tpu.RunConfig`- every single example I've been able to find sets `evaluation_master` to the same value as `master`, or leaves it unset.

- I've seen some mentions, circa September-November 2018, that `train_and_evaluate` is unsupported for TPUs, but almost nothing afterward as to whether or not that's still the case. Even if `train_and_evaluate` doesn't work as expected today, I'm still curious about nontrivial examples for `evaluation_master`.

Is there anything simple that I'm doing wrong? If my approach is wrong at a high level, how should I go about architecting my pipeline so to get the high-level behavior I listed at the beginning (i.e. one train node, one eval node, eval does periodic evaluations during training).

For my own amusement, why is this not a more common use case? I feel like seeing loss during training to get a sense of overfitting, and also not wanting the slowdown of graph creation/tear down is a no-brainer.

Lastly, apologies if this is not the right place for this question.  If that's the case, please let me know what might be a better space.

Thank you for reading through this!

Addendum: The use case here is fine-tuning BERT.

**Comment by sorenbouma** [comment]
> For my own amusement, why is this not a more common use case? I feel like seeing loss during training to get a sense of overfitting, and also not wanting the slowdown of graph creation/tear down is a no-brainer.

I've been running a similar workflow - just run `train` on a TPU pod and  run `evaluate` on another TPU with the same bucket specified in `model_dir` whenever a new checkpoint is saved.

+1 that this is a common need - seeing validation loss during training can be very important when working with non standard datasets/models, and rebuilding the graph every time to evaluate is slow.

It would be nice to be able to set up an eval worker on TPU through cluster configuration rather than running the script twice with different flags.

**Comment by vlfgns5** [comment]
I closed an issue by error when referencing this issue.

**Comment by sailordiary** [comment]
Came across this issue by chance. [This](https://github.com/google-research/ALBERT/blob/master/run_pretraining.py#L550) might be a workaround -- in the ALBERT codebase they run a continuous evaluation daemon loop, checking for a `estimator.latest_checkpoint()` every second until the training session reaches the maxmium number of iterations.
